osu.Game/Modes/LegacyReplay.cs

```diff
@@ -0,0 +1,274 @@
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using osu.Framework.Input;
+using osu.Framework.MathUtils;
+using osu.Game.Input.Handlers;
+using osu.Game.IO.Legacy;
+using OpenTK;
+using OpenTK.Input;
+using KeyboardState = osu.Framework.Input.KeyboardState;
+using MouseState = osu.Framework.Input.MouseState;
+
+namespace osu.Game.Modes
+{
+    public class LegacyReplay : Replay
+    {
+        protected List<LegacyReplayFrame> Frames = new List<LegacyReplayFrame>();
+
+        protected LegacyReplay()
+        {
+
+        }
+
+        public LegacyReplay(StreamReader reader)
+        {
+            float lastTime = 0;
+
+            foreach (var l in reader.ReadToEnd().Split(','))
+            {
+                var split = l.Split('|');
+
+                if (split.Length < 4 || float.Parse(split[0]) < 0) continue;
+
+                lastTime += float.Parse(split[0]);
+
+                Frames.Add(new LegacyReplayFrame(
+                    lastTime,
+                    float.Parse(split[1]),
+                    384 - float.Parse(split[2]),
+                    (LegacyButtonState)int.Parse(split[3])
+                    ));
+            }
+        }
+
+        public override ReplayInputHandler GetInputHandler() => new LegacyReplayInputHandler(Frames);
+
+        /// <summary>
+        /// The ReplayHandler will take a replay and handle the propagation of updates to the input stack.
+        /// It handles logic of any frames which *must* be executed.
+        /// </summary>
+        public class LegacyReplayInputHandler : ReplayInputHandler
+        {
+            private readonly List<LegacyReplayFrame> replayContent;
+            int currentFrameIndex;
+
+            public LegacyReplayFrame CurrentFrame => !hasFrames ? null : replayContent[currentFrameIndex];
+            public LegacyReplayFrame NextFrame => !hasFrames ? null : replayContent[MathHelper.Clamp(currentDirection > 0 ? currentFrameIndex + 1 : currentFrameIndex - 1, 0, replayContent.Count - 1)];
+
+            public LegacyReplayInputHandler(List<LegacyReplayFrame> replayContent)
+            {
+                this.replayContent = replayContent;
+            }
+
+            private bool nextFrame()
+            {
+                int newFrame = MathHelper.Clamp(currentFrameIndex + (currentDirection > 0 ? 1 : -1), 0, replayContent.Count - 1);
+
+                //ensure we aren't at an extent.
+                if (newFrame == currentFrameIndex) return false;
+
+                currentFrameIndex = newFrame;
+                return true;
+            }
+
+            public void SetPosition(Vector2 pos)
+            {
+            }
+
+            private Vector2? position
+            {
+                get
+                {
+                    if (!hasFrames)
+                        return null;
+
+                    if (AtLastFrame)

```

Comment:
NextFrame is already clamped, so the interpolation should give you the correct result (it interpolates between the same 2 values). No need for additional branches.